core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -1916,7 +1908,6 @@ protected void doRun() throws Exception {
 
         @Override
         protected void handleMergeException(final Directory dir, final Throwable exc) {
-            logger.error("failed to merge", exc);
```

Comment:
nit: shall we keep this here, for fast visibility and also, just in case it happens during node shutdown..

Reply:
I glanecd at the lucene code leading here and it seems all good. I do wonder if it's good to assert we can't find a Error in the cause chain of exc.

Reply:
Looking at the Lucene code as written right now, that's not possible. I'm really not sure if I want to add this assertion right now (look at how complicated it is in `Netty4Utils#maybeError` where Netty can hide things we do not want hidden from us). I understand the point of the assertion is to ensure that what we think we see in the Lucene code always holds, but then it's not clear to me if we'd ever encounter a situation where an `Error` occurs in testing that would trip one of these assertions anyway. So I see a cost to adding the assertion, and I'm not sure if I see benefit more than that cost.

Reply:
I see your point about tests maybe not being enough. I was thinking of something simple like `assert ExceptionsHelper.unwrap(exc, Error.class) == null : exc` .  the maybeError also looks for suppressions which I'm fine with not doing. Alternatively we can convert maybError to a top level utility.

Reply:
I will consider this in a follow-up.